Solutions/1.1-Introduction-to-Schnorr-Solutions.ipynb

```diff
@@ -0,0 +1,423 @@
+    "\n",
+    "Consider: `(a - b) * P = a * G - b * G`\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "a = random.randrange(1, SECP256K1_ORDER)\n",
+    "b = random.randrange(1, SECP256K1_ORDER)\n",
+    "\n",
+    "# Left: Compute a - b\n",
+    "a_neg_b = (a + (SECP256K1_ORDER - b)) % SECP256K1_ORDER\n",
+    "\n",
+    "# Left: Compute (a - b) * G\n",
+    "a_neg_b_key = ECKey()\n",
+    "a_neg_b_key.set(a_neg_b.to_bytes(32,'big'), True)      \n",
+    "Pk_left = a_neg_b_key.get_pubkey()\n",
+    "\n",
+    "# Right: Compute aG\n",
+    "a_key = ECKey()\n",
+    "a_key.set(a.to_bytes(32,'big'), True)      \n",
+    "Pk_a = a_key.get_pubkey()\n",
+    "\n",
+    "# Right: Compute -bG\n",
+    "# Use int() for -b\n",
+    "b_neg = SECP256K1_ORDER - b\n",
+    "b_neg_key = ECKey()\n",
+    "b_neg_key.set(b_neg.to_bytes(32,'big'), True)      \n",
+    "Pk_neg_b = b_neg_key.get_pubkey()\n",
+    "\n",
+    "# Right: Compute aG - bG\n",
+    "Pk_right = Pk_a.add(Pk_neg_b)\n",
+    "\n",
+    "# Left/Right: Assert equality\n",
+    "print(Pk_left.get_bytes() == Pk_right.get_bytes())\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 2: Schnorr Signatures\n",
+    "\n",
+    "The Schnorr signature equation is the following:\n",
+    "\n",
+    "* `S = R + H(x(R)|P|msg) * P`\n",
+    "\n",
+    "Signing involves generating a secret nonce first.\n",
+    "\n",
+    "* Generate secret scalar `k`\n",
+    "\n",
+    "Then computing s from:\n",
+    "\n",
+    "* `s = k - H(x(R)|P|m)`\n",
+    "\n",
+    "The resulting signature is:\n",
+    "\n",
+    "* `x(R), s`\n",
+    "\n",
+    "\n",
+    "\n",
+    "![test](../images/schnorr2.jpg)\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Constraint on the private nonce k.\n",
+    "\n",
+    "The Schnorr signature provides the verifier with the x-coordinate of the Nonce point R and s value. For a given x-coordinate on the SECP251K1 curve, there are two possible curve points:\n",
+    "\n",
+    "* `y^2 = x^3 + 7` (Two y-coordinate values for a given x-coordinate)\n",
+    "    * For x, both P(x,y) and P(x,-y) are valid curve points.\n",
+    "    * One of the y-coordinates is even, and the other one is odd.\n",
+    "\n",
+    "The nonce k is therefore constrained by requiring it to be a quadratic residue of p (SECP256k1 curve field size). \n",
+    "\n",
+    "* `k` and its complement `SECP26k1_order - k` will have nonce points `R(x, y)` and `R(x, -y)`.\n",
+    "    * Only one will be a quadratic residue of the curve order.\n",
+    "    \n",
+    "Whether a scalar is a quadratic residue of the secp256k1 field size is determined with the jacobi symbol:\n",
+    "\n",
+    "* Bitcoin Core method: `jacobi_symbol(k, SECP256K1_FIELD_SIZE) == 1`\n",
```

Comment:
the Jacobi symbol isn't tested on `k`. it's tested on the Y of kG.
we want to know that kG produces a quad residue Y.

Reply:
fixed